zebra-consensus/src/checkpoint/types.rs

```diff
@@ -33,7 +43,10 @@ impl Ord for Progress<BlockHeight> {
             (_, BeforeGenesis) => Ordering::Greater,
             (FinalCheckpoint, _) => Ordering::Greater,
             (_, FinalCheckpoint) => Ordering::Less,
-            (PreviousCheckpoint(self_height), PreviousCheckpoint(other_height)) => {
+            (InitialTip(self_height), InitialTip(other_height))
+            | (InitialTip(self_height), PreviousCheckpoint(other_height))
+            | (PreviousCheckpoint(self_height), InitialTip(other_height))
+            | (PreviousCheckpoint(self_height), PreviousCheckpoint(other_height)) => {
```

Comment:
wow, that's definitely a pattern!

I wonder if it would also be valid to write it like this

`(InitialTip(self_height) | PreviousCheckpoint(self_height), InitialTip(other_height) | PreviousCheckpoint(other_height)) => ...`

Edit: GAH, it's almost valid but still experimental

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=249eede9b3e9324b6a0d22abda7bb3de

https://github.com/rust-lang/rust/issues/54883

some day we will have all of the nice things :/